cpp/references/data_structures/list_digraph.cpp

```diff
@@ -3,636 +3,570 @@
-	Vector<ArcID> ret;
-	for (LemonArcIt it(_graph); it != lemon::INVALID; ++it) {
-		ret.push_back(LemonDigraph::id(it));
+		_graph.erase(handle);
 	}
-	return ret;
 }
-
-void ListDigraph::clear() {
-	_graph.clear();
+void ListDigraph::_erase_arcs(Variant p_arcs) {
+	ERR_FAIL_COND(p_arcs.get_type() != Variant::ARRAY);
+	erase_arcs(p_arcs);
 }
 
-void ListDigraph::set_data(ListDigraph::GraphComponent p_type, int64_t p_id, Variant p_data) {
-	switch (p_type)
-	{
-	case ListDigraph::VERTEX:
-	{
-		LemonNode node = LemonDigraph::nodeFromId(p_id);
-		ERR_FAIL_COND(!_graph.valid(node));
-		_vertex_map.set(node, p_data);
-		return;
-	}
-	case ListDigraph::ARC:
-	{
-		LemonArc arc = LemonDigraph::arcFromId(p_id);
-		ERR_FAIL_COND(!_graph.valid(arc));
-		_arc_map.set(arc, p_data);
-		return;
-	}
-	default:
-	{
-		ERR_FAIL();
-		break;
-	}
-	}
+void ListDigraph::_reserve_arcs(Variant p_amount) {
+	ERR_FAIL_COND(p_amount.get_type() != Variant::INT);
+	reserve_arcs(p_amount);
 }
 
-Variant ListDigraph::get_data(ListDigraph::GraphComponent p_type, int64_t p_id) {
-	switch (p_type)
-	{
-	case ListDigraph::VERTEX:
-	{
-		LemonNode node = LemonDigraph::nodeFromId(p_id);
-		ERR_FAIL_COND_V(!_graph.valid(node), Variant());
-		return _vertex_map.operator[](node);
-	}
-	case ListDigraph::ARC:
-	{
-		LemonArc arc = LemonDigraph::arcFromId(p_id);
-		ERR_FAIL_COND_V(!_graph.valid(arc), Variant());
-		return _arc_map.operator[](arc);
-	}
-	default:
-	{
-		ERR_FAIL();
-		break;
-	}
+Array ListDigraph::_get_arcs() {
+	Array ret;
+	for (LemonDigraph::ArcIt it(_graph); it != lemon::INVALID; ++it) {
+		ret.append(get_instance(it));
 	}
+	return ret;
 }
 
-void ListDigraph::update_data(ListDigraph::GraphComponent p_type, const HashMap<int64_t, Variant> &p_data) {
-	HashMap<int64_t, Variant>::const_iterator an_id_to_value;
-	switch (p_type)
-	{
-	case ListDigraph::VERTEX:
-		for (an_id_to_value = p_data.begin(); an_id_to_value != p_data.end(); ++an_id_to_value) {
-			
-			LemonNode node = LemonDigraph::nodeFromId(an_id_to_value->first);
-			ERR_CONTINUE(!_graph.valid(node));
-			_vertex_map.set(node, an_id_to_value->second);
+void ListDigraph::update_data(const Dictionary &p_data) {
+	Array refs = p_data.keys();
+	for (int i = 0; i < refs.size(); ++i) {
+		Variant ref = refs[i];
+		ERR_FAIL_COND(ref.get_type() != Variant::OBJECT);
+
+		Ref<Vertex> vertex = ref;
+		if (vertex.is_valid()) {
+			vertex->set_data(p_data[ref]);
+			continue;
 		}
-		break;
-	case ListDigraph::ARC:
-		for (an_id_to_value = p_data.begin(); an_id_to_value != p_data.end(); ++an_id_to_value) {
-			LemonArc arc = LemonDigraph::arcFromId(an_id_to_value->first);
-			ERR_CONTINUE(!_graph.valid(arc));
-			_arc_map.set(arc, an_id_to_value->second);
+
+		Ref<Arc> arc = ref;
+		if (arc.is_valid()) {
+			arc->set_data(p_data[ref]);
+			continue;
 		}
-		break;
-	default:
-		break;
+
+		ERR_CONTINUE(!arc.is_valid() || !vertex.is_valid());
 	}
 }
+void ListDigraph::_update_data(Variant p_data) {
```

Comment:
I know we discussed removing the C++-implemented, scripting-API-compatible methods and instead just putting the Variant wrapper / type check inside the Variant-accepting method instead (so, e.g. putting the `_update_data` logic inside the `update_data` method and binding to the `update_data` method instead). Did we only discuss it and never decide to do it? I think it's probably a good idea, but what do you think?

Reply:
I haven't done it yet... was waiting for your full review. should I do it?

Reply:
Sure.